spritzle/resume_data.py

```diff
@@ -0,0 +1,128 @@
+#
+# spritzle/resume_data.py
+#
+# Copyright (C) 2016 Andrew Resch <andrewresch@gmail.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.    See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.    If not, write to:
+#   The Free Software Foundation, Inc.,
+#   51 Franklin Street, Fifth Floor
+#   Boston, MA    02110-1301, USA.
+#
+
+import asyncio
+import binascii
+import functools
+import logging
+from pathlib import Path
+
+import libtorrent as lt
+
+log = logging.getLogger('spritzle')
+
+
+class ResumeData(object):
+    def __init__(self, core):
+        self.core = core
+        self.loop = asyncio.get_event_loop()
+        self.save_loop_task = None
+
+        # Store state of outstanding save resume data alerts
+        self.resume_data_futures = {}
+
+    async def start(self):
+        log.debug('Resume data manager starting...')
+        await self.load()
+        self.core.alert.register_handler(
+            'save_resume_data_alert',
+            self.on_save_resume_data_alert
+        )
+        self.core.alert.register_handler(
+            'save_resume_data_failed_alert',
+            self.on_save_resume_data_failed_alert
+        )
+        self.save_loop_task = asyncio.ensure_future(self.save_loop())
+
+    async def stop(self):
+        log.debug("Resume data manager stopping...")
+        if self.save_loop_task:
+            self.save_loop_task.cancel()
+            await self.save_loop_task
+        await self.save()
+        log.debug("Resume data manager stopped.")
+
+    async def save_loop(self):
+        save_fut = None
+        try:
+            while True:
+                await asyncio.sleep(
+                    self.core.config['resume_data_save_frequency'])
+                # Don't interrupt save process when loop is cancelled
+                save_fut = asyncio.ensure_future(self.save())
+                await asyncio.shield(save_fut)
+        except asyncio.CancelledError:
+            if save_fut and not save_fut.done():
+                await save_fut
+
+    async def on_save_resume_data_alert(self, alert):
```

Comment:
We could test these by:
* adding the torrent to the session
* calling this function with a mock alert
* assert that the file was created with some data in it, maybe bdecode it and look for a key we know should be there

Reply:
Isn't that sort of already happening in the 'test_save' test? (except instead of a mocked alert, a real one comes in.) I agree on adding a bit to bdecode the data and test it's like what we expect though.